### DISCUSSION TITLE
Additional Range of Motion Constraint for GenericJoint

### DISCUSSION THREAD

**Discussion by csas8682** [jgerstmayr/EXUDYN#17] [Q&A on usage of Exudyn] [answered]
Hello, 

I was wondering if it would be possible, to set a further constraint to a joint in _Exudyn_. 
In addition to constrain an axis, I would need to allow the movement only in a certain range of an axis (e. g. 0° - 180°) anything, more than 180° should be constrained as well. 
I was looking for a _UserFunction_ in _TheDoc_, I only found an _offsetUserFunction_ for GenericJoints, but as far as I understood it would not work for me problem. 

Once again, thank you in advance for your help! 🙏

**Comment by jgerstmayr** [comment, marked as answer]
This is an interesting question, and it is in general possible. In a recent test, which did not yet make it into the examples section of Exudyn, we did the following for a TorsionalSpringDamper (which could be also adapted for a regular SpringDamper):

def UFtsd(mbs, t, itemNumber, rotation, angularVelocity, stiffness, damping, offset):
    f = 0.
    limPos=pi
    limNeg=-pi
    if rotation > limPos:
        f = 50*stiffness*(rotation-limPos)**2 + stiffness*(rotation-offset) + damping*angularVelocity
    elif rotation < limNeg:
        f = -50*stiffness*(rotation-limNeg)**2 + stiffness*(rotation-offset) + damping*angularVelocity
    else:
        f = stiffness*(rotation-offset) + damping*angularVelocity
    return f

This limits our joint with a quadratically increasing stiffness. In future, we may add such functionality to torsional and translational spring dampers, which will increase the performance as compared to user functions.
